src/core/common/time.hpp

```diff
@@ -0,0 +1,359 @@
+/*
+ *  Copyright (c) 2019, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for time instance.
+ */
+
+#ifndef TIME_HPP_
+#define TIME_HPP_
+
+#include "openthread-core-config.h"
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <openthread/platform/alarm-micro.h>
+#include <openthread/platform/alarm-milli.h>
+
+namespace ot {
+
+/**
+ * @addtogroup core-timer
+ *
+ * @brief
+ *   This module includes definitions for the time instance.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This class represent a time instance.
+ *
+ */
+class Time
+{
+public:
+    /**
+     * This type represents a time duration.
+     *
+     */
+    typedef uint32_t Duration;
+
+    /**
+     * This type represents the underlying unsigned int type used to store numeric value of a `Time` instance.
+     *
+     */
+    typedef uint32_t Value;
+
+    /**
+     * This constant defines a forever duration ensured to be longer than any other duration.
+     *
+     */
+    static const Duration kForeverDuration = ~static_cast<Duration>(0UL);
```

Comment:
Is this really "forever"? Is `kMaxDuration` more appropriate?

Reply:
We have a `kMaxDuration` definition (in `Timer` class) which is the max duration we can use in a timer APIs `Start()` and `StartAt()`. Even though we can have a `Time:kMaxDuration` and a `Timer::kMaxDuration` I think it'd be better to use a different name for these two definitions.  I opted for `kForeverDuration` which was similar to `kForeverDt` we were using before. Suggestions for new names for these are most welcome. :)

Reply:
How about `Timer::kMaxDelay` vs. `Time::kMaxDuration`?